provider/core/src/request.rs

```diff
@@ -481,31 +656,207 @@ impl DataLocale {
+    ///
+    /// For more information and examples, see [`AuxiliaryKey`].
+    pub fn has_aux(&self) -> bool {
+        self.aux.is_some()
+    }
+
+    /// Sets an auxiliary key on this [`DataLocale`].
+    ///
+    /// Returns the previous auxiliary key if present.
+    ///
+    /// For more information and examples, see [`AuxiliaryKey`].
+    pub fn set_aux(&mut self, value: AuxiliaryKey) -> Option<AuxiliaryKey> {
+        self.aux.replace(value)
+    }
+
+    /// Remove an auxiliary key, if present. Returns the removed auxiliary key.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_locid::locale;
+    /// use icu_provider::prelude::*;
+    /// use writeable::assert_writeable_eq;
+    ///
+    /// let mut data_locale: DataLocale = locale!("ar-EG").into();
+    /// let aux = "GBP".parse::<AuxiliaryKey>().expect("contains valid characters");
+    /// data_locale.set_aux(aux);
+    /// assert_writeable_eq!(data_locale, "ar-EG$GBP");
+    ///
+    /// let maybe_aux = data_locale.remove_aux();
+    /// assert_writeable_eq!(data_locale, "ar-EG");
+    /// assert_writeable_eq!(maybe_aux.unwrap(), "GBP");
+    /// ```
+    pub fn remove_aux(&mut self) -> Option<AuxiliaryKey> {
+        self.aux.take()
+    }
+}
+
+/// The "auxiliary key" is an annotation on [`DataLocale`] that can contain an arbitrary
+/// information that does not fit into the [`LanguageIdentifier`] or [`Keywords`].
+///
+/// It is represented as a string separated from the BCP-47 ID with a `$`.
+///
+/// An auxiliary key currently allows alphanumerics and `-`.
+///
+/// # Examples
+///
+/// ```
+/// use icu_locid::locale;
+/// use icu_provider::prelude::*;
+/// use writeable::assert_writeable_eq;
+///
+/// let mut data_locale: DataLocale = locale!("ar-EG").into();
+/// assert_writeable_eq!(data_locale, "ar-EG");
+/// assert!(!data_locale.has_aux());
+/// assert_eq!(data_locale.get_aux(), None);
+///
+/// let aux = "GBP".parse::<AuxiliaryKey>().expect("contains valid characters");
+///
+/// data_locale.set_aux(aux);
+/// assert_writeable_eq!(data_locale, "ar-EG$GBP");
+/// assert!(data_locale.has_aux());
+/// assert_eq!(data_locale.get_aux(), Some(&"GBP".parse().unwrap()));
+/// ```
+///
+/// Not all strings are valid auxiliary keys:
+///
+/// ```
+/// use icu_provider::prelude::*;
+///
+/// assert!("abcdefg".parse::<AuxiliaryKey>().is_ok());
+/// assert!("ABC123".parse::<AuxiliaryKey>().is_ok());
+/// assert!("abc-xyz".parse::<AuxiliaryKey>().is_ok());
+///
+/// assert!("".parse::<AuxiliaryKey>().is_err());
+/// assert!("!@#$%".parse::<AuxiliaryKey>().is_err());
+/// assert!("abc_xyz".parse::<AuxiliaryKey>().is_err());
+/// assert!("abc$xyz".parse::<AuxiliaryKey>().is_err());
+/// ```
+///
+/// [`Keywords`]: unicode_ext::Keywords
+#[derive(Debug, PartialEq, Clone, Default, Eq, Hash)]
+pub struct AuxiliaryKey {
+    // DISCUSS: SmallStr? TinyStrAuto?
+    // DISCUSS: Make this a dynamically sized type so references can be taken?
+    value: String,
```

Comment:
I think this should be `Cow<str>` to allow const-constructing with a `data_locale!` macro. Also resizing doesn't seem to be use case, so there's no need for separate length and capacity.

Reply:
Note that `Cow<str>` uses a `String` in the owned variant.

How about something like

```rust
enum BoxOrStatic {
    Box(Box<str>),
    Static(&'static str),
    Tiny(TinyAsciiStr<23>),
}
```

The size of that is 24 bytes, which is the same as a `String`.

https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=5d7c515c26f12e820c932959599bfb56

The main downside is that code which gets a `&str` from that thing is going to need to do more branching.

Reply:
Ah it does, I somehow thought it'd be `Box<str>`. That enum looks good.

Reply:
@Manishearth Opinions on the enum before I spend time implementing it? I'm a little concerned about the branching and resulting code size but I haven't measured it yet.